### PR TITLE
fix ts for long double and wchar

### PR DESCRIPTION
**Pull request by dirk-thomas**
Same as ros2/rosidl#383.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Including the second commit addressing `wchar` support:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7295/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3470/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5970/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7115/)

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 1]
```diff
@@ -348,6 +358,12 @@ else:
       cdr >> tmp;
       array_ptr[i] = tmp ? true : false;
     }
+@[    elif isinstance(member.type.value_type, BasicType) and member.type.value_type.typename == 'wchar']@
+    for (size_t i = 0; i < size; ++i) {
+      wchar_t tmp;
+      cdr >> tmp;
+      array_ptr[i] = static_cast<char16_t>(tmp);
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: @dirk-thomas hmm, `wchar_t` is not guaranteed to be 16 bits wide but I presume Fast-RTPS always deserializes UTF-16, is that correct?

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 2]
```diff
@@ -213,6 +213,14 @@ if isinstance(type_, AbstractNestedType):
       rosidl_typesupport_fastrtps_c::u16string_to_wstring(*str, wstr);
       cdr << wstr;
     }
+@[    elif isinstance(member.type.value_type, BasicType) and member.type.value_type.typename == 'wchar']@
+    for (size_t i = 0; i < size; ++i) {
+      if (!callbacks->cdr_serialize(
+          static_cast<char16_t *>(&array_ptr[i]), cdr))
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: @dirk-thomas isn't it a `char16_t *` already? should this be `wchar_t` instead?

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 1 above)
Comment (dirk-thomas) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: On the wire FastRTPS uses 16 bit. It is agnostic to the actual encoding and only treats it as binary data.

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 2 above)
Comment (dirk-thomas) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: Great catch. That is definitely wrong as is: fixed in 73e975baffa3361098c99f3b6298cef58bba8d8c.

**Pull request merged**